typing/ctype.ml

```diff
@@ -4520,47 +4678,71 @@ let cyclic_abbrev env id ty =
   in check_cycle [] ty
 
 (* Check for non-generalizable type variables *)
-exception Non_closed0
+exception Nongen
 let visited = ref TypeSet.empty
 
-let rec closed_schema_rec env ty =
+let rec nongen_schema_rec env ty =
   let ty = repr ty in
   if TypeSet.mem ty !visited then () else begin
     visited := TypeSet.add ty !visited;
     match ty.desc with
       Tvar _ when ty.level <> generic_level ->
-        raise Non_closed0
+        raise Nongen
     | Tconstr _ ->
         let old = !visited in
-        begin try iter_type_expr (closed_schema_rec env) ty
-        with Non_closed0 -> try
+        begin try iter_type_expr (nongen_schema_rec env) ty
+        with Nongen -> try
           visited := old;
-          closed_schema_rec env (try_expand_head try_expand_safe env ty)
+          nongen_schema_rec env (try_expand_head try_expand_safe env ty)
         with Cannot_expand ->
-          raise Non_closed0
+          raise Nongen
         end
     | Tfield(_, kind, t1, t2) ->
         if field_kind_repr kind = Fpresent then
-          closed_schema_rec env t1;
-        closed_schema_rec env t2
+          nongen_schema_rec env t1;
+        nongen_schema_rec env t2
     | Tvariant row ->
         let row = row_repr row in
-        iter_row (closed_schema_rec env) row;
-        if not (static_row row) then closed_schema_rec env row.row_more
+        iter_row (nongen_schema_rec env) row;
+        if not (static_row row) then nongen_schema_rec env row.row_more
     | _ ->
-        iter_type_expr (closed_schema_rec env) ty
+        iter_type_expr (nongen_schema_rec env) ty
   end
 
 (* Return whether all variables of type [ty] are generic. *)
-let closed_schema env ty =
+let nongen_schema env ty =
   visited := TypeSet.empty;
   try
-    closed_schema_rec env ty;
-    visited := TypeSet.empty;
-    true
-  with Non_closed0 ->
+    nongen_schema_rec env ty;
     visited := TypeSet.empty;
     false
+  with Nongen ->
+    visited := TypeSet.empty;
+    true
+
+(* Check that all type variables are generalizable *)
+(* Use Env.empty to prevent expansion of recursively defined object types;
+   cf. typing-poly/poly.ml *)
+let rec nongen_class_type = function
+  | Cty_constr (_, params, _) ->
+      List.exists (nongen_schema Env.empty) params
+  | Cty_signature sign ->
+      nongen_schema Env.empty sign.csig_self
+      || nongen_schema Env.empty sign.csig_self_row
+      || Meths.exists
+           (fun _ (_, _, ty) -> nongen_schema Env.empty ty)
+           sign.csig_meths
```

Comment:
The check on the row type should always be redundant since it is within the self type, but I prefer to include it and have the code obviously correct rather than subtly correct. The check on self is redundant when we have called `update_class_signature`, which we will always have done when we call this function, but again I'd rather leave the redundancy and not rely on that.